### ISSUE TITLE
Reason for the proposed optimization method 

### ISSUE THREAD

**Issue by puzzlecollector** [stanford-futuredata/ColBERT#123]
From what I understand, the loss for ColBERT is calculated is follows: 
score = some metric, like L2-norm or cosine similarity 
q,p,n are query, positive and negative respectively 

score(q, p), score(q, n) is calculated then we apply softmax then use the ground truth label (always zero) to compute the cross entropy loss. 

Did the authors use this method over other losses (such as TripletMarginLoss) because they thought this method of training is more effective?

**Comment by r1ckC139** [comment]
Hi @puzzlecollector,
I see that you have a lot of questions and inquiries about ColBERT, and I am assuming that you have a very good understanding of how ColBERT works and this Repo, so I want to ask you a question (I'm so sorry for asking you here if you mind please ignore this comment, but the reason I ask you instead of creating a new issue is that they are usually too slow to respond), anyway my question is the step re-ranking after doing the retrieval step (unorder.tsv file), I really don't understand what this rank function (picture and link below) does, even though I tried hard to figure it out, it has so many variables that I'm really confused. If you understand this function, hope you can explain it to me. Thank you very much (Once again really sorry for asking it here) !!!
[Rank Function](https://github.com/jonsaadfalcon/ColBERT/blob/b492982b3192fe5b2e8fb0d490c86f4842a195ef/colbert/ranking/index_ranker.py#L56)
![image](https://user-images.githubusercontent.com/61724310/183598377-b19dc667-8aa5-4ada-a496-c37693289b25.png)

**Comment by okhat** [comment]
It's fairly standard.
